tomcat/tomcat-jta/src/main/java/org/jboss/narayana/tomcat/jta/TransactionalDataSourceFactory.java

```diff
@@ -0,0 +1,183 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.jboss.narayana.tomcat.jta;
+
+import com.arjuna.ats.internal.jta.recovery.arjunacore.XARecoveryModule;
+import com.arjuna.ats.jta.recovery.XAResourceRecoveryHelper;
+import org.apache.commons.dbcp2.PoolableConnection;
+import org.apache.commons.dbcp2.PoolableConnectionFactory;
+import org.apache.commons.dbcp2.managed.DataSourceXAConnectionFactory;
+import org.apache.commons.dbcp2.managed.ManagedDataSource;
+import org.apache.commons.pool2.impl.GenericObjectPool;
+import org.apache.commons.pool2.impl.GenericObjectPoolConfig;
+
+import javax.naming.Context;
+import javax.naming.Name;
+import javax.naming.RefAddr;
+import javax.naming.Reference;
+import javax.naming.spi.ObjectFactory;
+import javax.sql.XADataSource;
+import javax.transaction.TransactionManager;
+import javax.transaction.xa.XAResource;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.Type;
+import java.sql.SQLException;
+import java.util.Hashtable;
+import java.util.Properties;
+
+/**
+ * @author <a href="mailto:zfeng@redhat.com>Zheng Feng</a>
+ */
+public class TransactionalDataSourceFactory implements ObjectFactory {
+    private static final String PROP_USERNAME = "username";
+    private static final String PROP_PASSWORD = "password";
+    private static final String PROP_MAX_TOTAL = "maxTotal";
+    private static final String PROP_MAX_IDLE = "maxIdle";
+    private static final String PROP_MIN_IDLE = "minIdle";
+
+    private static final String[] ALL_PROPERTIES = {
+            PROP_USERNAME,
+            PROP_PASSWORD,
+            PROP_MAX_TOTAL,
+            PROP_MAX_IDLE,
+            PROP_MIN_IDLE
+    };
+
+    @Override
+    public Object getObjectInstance(Object obj, Name name, Context context, Hashtable<?, ?> environment) throws Exception {
+
+        if (obj == null || !(obj instanceof Reference)) {
+            return null;
+        }
+
+        final Reference ref = (Reference) obj;
+        if (!"javax.sql.DataSource".equals(ref.getClassName())) {
+            return null;
+        }
+
+        final Properties properties = new Properties();
+        for (final String propertyName : ALL_PROPERTIES) {
+            final RefAddr ra = ref.get(propertyName);
+            if (ra != null) {
+                properties.setProperty(propertyName, ra.getContent().toString());
+            }
+        }
+
+        TransactionManager transactionManager = (TransactionManager) getReferenceObject(ref, context, "transactionManager");
+        XADataSource xaDataSource = (XADataSource) getReferenceObject(ref, context, "xaDataSource");
```

Comment:
What sets the requirement for xaDataSource - perhaps we should add this to our docs?

Reply:
Or at least a comment to where it comes from - I guess it might be the attribute in the resource in the context.xml?